src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,488 @@
+using System.Security.Claims;
+using System.Security.Principal;
+using GraphQL.Execution;
+using GraphQL.Transport;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+
+namespace GraphQL.Server.Transports.AspNetCore.WebSockets;
+
+/// <summary>
+/// Manages a WebSocket message stream.
+/// </summary>
+public abstract partial class BaseSubscriptionServer : IOperationMessageProcessor
+{
+    private volatile int _initialized;
+    private CancellationTokenSource? _cancellationTokenSource;
+    private readonly GraphQLHttpMiddlewareOptions _options;
+
+    /// <summary>
+    /// Returns a <see cref="IWebSocketConnection"/> instance that can be used
+    /// to send messages to the client.
+    /// </summary>
+    protected IWebSocketConnection Connection { get; }
+
+    /// <summary>
+    /// Returns a <see cref="System.Threading.CancellationToken"/> that is signaled
+    /// when this class is disposed or when the WebSockets connection is closed.
+    /// Intended to stop background timers.
+    /// </summary>
+    protected CancellationToken CancellationToken { get; }
+
+    /// <summary>
+    /// Returns a synchronized list of subscriptions.
+    /// </summary>
+    protected SubscriptionList Subscriptions { get; } = new();
+
+    /// <summary>
+    /// Returns the default keep-alive timeout.
+    /// </summary>
+    protected virtual TimeSpan DefaultKeepAliveTimeout { get; } = Timeout.InfiniteTimeSpan;
+
+    /// <summary>
+    /// Returns the default connection timeout.
+    /// </summary>
+    protected virtual TimeSpan DefaultConnectionTimeout { get; } = TimeSpan.FromSeconds(10);
+
+    /// <summary>
+    /// Initailizes a new instance with the specified parameters.
+    /// </summary>
+    /// <param name="sendStream">The WebSockets stream used to send data packets or close the connection.</param>
+    /// <param name="options">Configuration options for this instance.</param>
+    public BaseSubscriptionServer(
+        IWebSocketConnection sendStream,
+        GraphQLHttpMiddlewareOptions options)
```

Comment:
`GraphQLHttpMiddlewareOptions` seems excessive here except its usage in `AuthorizeAsync` (3 AuthXXX properties).  In ideal subscription server should accept `GraphQLWebSocketOptions`.

Reply:
🤔 Let's consider such option - imagine separate auth settings for "usual" http middleware and for WebSocket transports. So copy 3 settings into `GraphQLWebSocketOptions`. Then add interface like `IAuthorizationOptions` with 3 settings and implement it by both `GraphQLHttpMiddlewareOptions` and `GraphQLWebSocketOptions`. Then change ` GraphQLHttpMiddlewareOptions middlewareOptions` to ` IAuthorizationOptions authorizationOptions` into `AuthorizationParameters`. Viola! Auth works for both transports independently. Even further - remove `WebSockets` property from `GraphQLHttpMiddlewareOptions` since now options are independent. `WebSocketConnection` can work with `GraphQLWebSocketOptions` directly. The same for `BaseSubscriptionServer`.

Reply:
My intent to not mix options for middleware and for websockets.

Reply:
I don’t want another copy of the auth settings as visible to the user. They are not logically separate settings. We can make a third options class and copy the applicable settings into it from GraphQLHttpMiddlewareOptions and GraphQLWebSocketOptions if you like.

Reply:
WebSocket is just another form of transport, just as POST is different than GET. Clients can request query or mutations via GET, POST or WebSocket. Within POST they can submit via multipart content type or json content type. None of that affects the authorization requirements.

Reply:
Another option is to have the WebSocket constructor accept both GraphQLWebSocketOptions and IAuthorizationOptions.

Reply:
> I don’t want another copy of the auth settings as visible to the user. They are not logically separate settings.

Why not? One can enable auth on one transport and disable on other. Or use policy for HTTP auth and roles for WebSocket.

![изображение](https://user-images.githubusercontent.com/21261007/172640100-3a83d636-819b-436e-a431-3496ebb018f0.png)
It's not true. I do not like that this options object is passed to those places where its properties are completely not needed and not applicable. `AuthorizationParameters` is only place that holds reference to `GraphQLHttpMiddlewareOptions` and forces `BaseSubscriptionServer` to use them.

Reply:
> I don’t want another copy of the auth settings as visible to the user. They are not logically separate settings. We can make a **third options class and copy the applicable settings into it** from GraphQLHttpMiddlewareOptions and GraphQLWebSocketOptions if you like.

I will not figure out how it looks.

Reply:
Sending query/mutation GraphQL requests over a WebSocket channel is not unlike a [HTTP persistent connection.](https://en.wikipedia.org/wiki/HTTP_persistent_connection)

Reply:
I think you should understand that one of my goals here is to make supporting WebSocket connections easy to use.  Currently you need to all of this:
1. Add another nuget package, which bring additional dependencies
2. Add two additional services in the graphql builder
3. Configure ASP.NET Core to enable websocket support
4. Add another line in the app builder to configure the websocket endpoint

With the new code:
- No additional nuget package is needed (and there are no additional dependencies)
- No additional configuration necessary in GraphQL builder
- Still need to enable ASP.NET Core websocket support
- No additional code in app builder

Basically, assuming you've enabled WebSockets through ASP.NET Core, it just works.  And if the authorization options are separate, I see a great danger that someone sets authorization options for the main middleware and forgets to set them in the websocket options as well.  Not only is it typically redundant, but this is a huge security concern.

There actually already exists an easy way to configure different security options for websocket transport.  Just add another line in the app builder.

```cs
app.UseGraphQL("/graphql", opts => {
    opts.HandleWebSockets = false;
    // set auth settings for get/post here
});
app.UseGraphQL("/graphql", opts => {
    opts.HandleGet = false;
    opts.HandlePost = false;
    // set auth settings for websockets here
});
```

(I wouldn't even document this, I see it as such a rare use case.)

But that is much better than the alternative scenario, where someone accidentally only sets auth options only for get/post and forgets to set them on websocket connections.

Reply:
> My intent is to not mix options for middleware and for websockets.

I'm open to resolve this in any way.